server/migrations/20220316164353-add-race-stats-to-matchmaking-ratings.js

```diff
@@ -0,0 +1,180 @@
+exports.up = async function (db) {
+  await db.runSql(`
+    ALTER TABLE matchmaking_ratings
+
+    ADD COLUMN p_wins integer NOT NULL DEFAULT 0,
+    ADD COLUMN p_losses integer NOT NULL DEFAULT 0,
+
+    ADD COLUMN t_wins integer NOT NULL DEFAULT 0,
+    ADD COLUMN t_losses integer NOT NULL DEFAULT 0,
+
+    ADD COLUMN z_wins integer NOT NULL DEFAULT 0,
+    ADD COLUMN z_losses integer NOT NULL DEFAULT 0,
+
+    ADD COLUMN r_wins integer NOT NULL DEFAULT 0,
+    ADD COLUMN r_losses integer NOT NULL DEFAULT 0,
+
+    ADD COLUMN r_p_wins integer NOT NULL DEFAULT 0,
+    ADD COLUMN r_p_losses integer NOT NULL DEFAULT 0,
+
+    ADD COLUMN r_t_wins integer NOT NULL DEFAULT 0,
+    ADD COLUMN r_t_losses integer NOT NULL DEFAULT 0,
+
+    ADD COLUMN r_z_wins integer NOT NULL DEFAULT 0,
+    ADD COLUMN r_z_losses integer NOT NULL DEFAULT 0;
+  `)
+
+  // Populate wins/losses for each race for each user
+  for (const race of ['p', 't', 'z', 'r']) {
+    await db.runSql(`
+      UPDATE matchmaking_ratings AS s
+      SET ${race}_wins = r.wins
+      FROM (
+        SELECT user_id, g.config->'gameSourceExtra'->>'type' as matchmaking_type, COUNT(*) as wins
+        FROM games_users gu JOIN games g
+          ON gu.game_id = g.id
+        WHERE
+          g.config->>'gameSource' = 'MATCHMAKING' AND
+          gu.selected_race = '${race}' AND
+          gu.result = 'win'
+        GROUP BY user_id, matchmaking_type
+      ) r
+      WHERE s.user_id = r.user_id AND s.matchmaking_type = r.matchmaking_type::matchmaking_type
+    `)
+
+    await db.runSql(`
+      UPDATE matchmaking_ratings AS s
+      SET ${race}_losses = r.losses
+      FROM (
+        SELECT user_id, g.config->'gameSourceExtra'->>'type' as matchmaking_type, COUNT(*) as losses
+        FROM games_users gu JOIN games g
+          ON gu.game_id = g.id
+        WHERE
+          g.config->>'gameSource' = 'MATCHMAKING' AND
+          gu.selected_race = '${race}' AND
+          gu.result = 'loss'
+        GROUP BY user_id, matchmaking_type
+      ) r
+      WHERE s.user_id = r.user_id AND s.matchmaking_type = r.matchmaking_type::matchmaking_type
+    `)
+  }
+
+  // Populate wins/losses for each assigned random race for each user
+  for (const race of ['p', 't', 'z']) {
+    await db.runSql(`
+      UPDATE matchmaking_ratings AS s
+      SET r_${race}_wins = r.wins
+      FROM (
+        SELECT user_id, g.config->'gameSourceExtra'->>'type' as matchmaking_type, COUNT(*) as wins
+        FROM games_users gu JOIN games g
+          ON gu.game_id = g.id
+        WHERE
+          g.config->>'gameSource' = 'MATCHMAKING' AND
+          gu.assigned_race = '${race}' AND
```

Comment:
I think you copied this from the previous migration, but this logic doesn't seem to be correct (and we may need to correct the previous migration as well). `assigned_race` is populated regardless of whether random was selected, and I don't see anything in this query that ensures it only applies to random, so adding e.g. `z_wins` and `r_z_wins` double-counts them

Reply:
If only I had caught it when I wrote the original code :P